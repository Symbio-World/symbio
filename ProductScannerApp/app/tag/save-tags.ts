import firestore from '@react-native-firebase/firestore'

type SaveTags = (userId: string, tags: string[]) => Promise<void>

export const command = 'saveTags'
export const saveTags: SaveTags = async (userId, tags) => {
  await firestore()
    .collection('events')
    .add({ userId, command: 'saveTags', tags })
}
