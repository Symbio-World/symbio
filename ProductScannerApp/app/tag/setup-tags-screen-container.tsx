import React, { useState, useEffect } from 'react'
import { useAsync } from 'react-async'
import { User } from '../auth'
import { SetupTagsScreen } from './setup-tags-screen'

type Props = {
  user: User
  onStore?: () => void
}

type Deps = {
  storeUserTags: (user: User, tags: string[]) => Promise<void>
  tags: string[]
}

type CreateSetupTagsScreenContainer = (deps: Deps) => React.FC<Props>

export const createSetupTagsScreenContainer: CreateSetupTagsScreenContainer = ({
  storeUserTags,
  tags,
}) => ({ user, onStore = () => {} }) => {
  const [selectedTags, setSelectedTags] = useState<string[]>([])
  const { isPending, error, run } = useAsync<void>({
    deferFn: ([user, tags]) => storeUserTags(user, tags),
  })

  const onTagPress = (tag: string) =>
    selectedTags.includes(tag)
      ? setSelectedTags(tags.filter(t => t !== tag))
      : setSelectedTags([tag, ...selectedTags])
  const handleSubmit = () => {
    run(user, selectedTags)
    onStore()
  }

  return (
    <SetupTagsScreen
      selectedTags={selectedTags}
      tags={tags}
      onTagPress={onTagPress}
      onSubmit={handleSubmit}
    />
  )
}
