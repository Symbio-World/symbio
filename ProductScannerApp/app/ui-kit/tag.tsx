import React from 'react'
import { Text, TouchableOpacity } from 'react-native'
import { t } from 'react-native-tailwindcss'

type Props = {
  title: string
  selected?: boolean
  onPress?: () => void
}
export const Tag: React.FC<Props> = ({
  title,
  onPress = () => {},
  selected = false,
}) => {
  return (
    <TouchableOpacity
      style={[
        selected ? t.bgGreen500 : t.bgWhite,
        t.roundedFull,
        t.m3,
        t.pY3,
        t.pX6,
      ]}
      onPress={onPress}
    >
      <Text
        style={[t.textXl, selected ? t.textWhite : t.textGreen500, t.fontBold]}
      >
        {title}
      </Text>
    </TouchableOpacity>
  )
}
