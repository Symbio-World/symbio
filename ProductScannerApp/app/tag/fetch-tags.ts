import firestore from '@react-native-firebase/firestore'
import { command } from './save-tags'

type FetchTags = (userId: string) => Promise<string[] | undefined>

export const fetchTags: FetchTags = async userId => {
  const snapshot = await firestore()
    .collection('events')
    .where('command', '==', command)
    .where('userId', '==', userId)
    .get()
  console.log(snapshot)
  return snapshot.docs.length > 0 ? snapshot.docs[0].data().tags : undefined
}
