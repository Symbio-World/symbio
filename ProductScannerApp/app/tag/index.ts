import firestore from '@react-native-firebase/firestore'
import { createSetupTagsScreenContainer } from './setup-tags-screen-container'

export const SetupTagsScreenContainer = createSetupTagsScreenContainer({
  storeUserTags: async ({ uid }, tags) => {
    await firestore()
      .collection('user-tags')
      .doc(uid)
      .set({ tags })
      .catch(console.log)
  },
  tags: [
    'Vegan',
    'Gluten free',
    'Nut free',
    'Lactose free',
    'Halal',
    'Organic',
    'Eco friendly',
  ],
})
