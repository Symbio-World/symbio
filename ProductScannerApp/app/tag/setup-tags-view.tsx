import React from 'react'
import { ScrollView, View, Text } from 'react-native'
import { t } from 'react-native-tailwindcss'
import { Tag } from '../ui-kit/tag'
import { Button } from '../ui-kit/button'

type Props = {
  tags?: string[]
  selectedTags?: string[]
  onTagPress?: (tag: string) => void
  onSubmit?: () => void
}
export const SetupTagsView: React.FC<Props> = ({
  tags = [],
  selectedTags = [],
  onTagPress = () => {},
  onSubmit = () => {},
}) => {
  return (
    <View style={[t.m5]}>
      <ScrollView>
        <View>
          <Text style={[t.text4xl, t.fontBold, t.mY4]}>ðŸ‘‹Hey There</Text>
          <Text style={[t.mY4]}>
            <Text style={[t.fontBold]}>Should I Buy</Text> lets you understand
            if the products you find in the supermarket fit your values.
          </Text>
          <Text style={[t.fontBold]}>What is important to you?</Text>
          <View style={[t.flexRow, t.flexWrap, t.mY4]}>
            {tags.map(t => (
              <Tag
                key={t}
                title={t}
                onPress={() => onTagPress(t)}
                selected={selectedTags.includes(t)}
              />
            ))}
          </View>
        </View>
        <View style={[t.h24]} />
      </ScrollView>
      <View style={[t.absolute, t.bottom0, t.wFull, t.bgWhite]}>
        <Button onPress={onSubmit} title='Get Started' />
      </View>
    </View>
  )
}
