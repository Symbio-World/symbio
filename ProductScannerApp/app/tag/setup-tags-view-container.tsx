import React, { useState } from 'react'
import { SetupTagsView } from './setup-tags-view'

type Props = {
  userId: string
  onStore?: () => void
}

type Deps = {
  saveTags: (userId: string, tags: string[]) => Promise<void>
  tags: string[]
}

type CreateSetupTagsViewContainer = (deps: Deps) => React.FC<Props>

export const createSetupTagsViewContainer: CreateSetupTagsViewContainer = ({
  saveTags,
  tags,
}) => ({ userId, onStore = () => {} }) => {
  const [selectedTags, setSelectedTags] = useState<string[]>([])

  const onTagPress = (tag: string) =>
    selectedTags.includes(tag)
      ? setSelectedTags(tags.filter(t => t !== tag))
      : setSelectedTags([tag, ...selectedTags])
  const handleSubmit = () => {
    saveTags(userId, selectedTags)
    onStore()
  }

  return (
    <SetupTagsView
      selectedTags={selectedTags}
      tags={tags}
      onTagPress={onTagPress}
      onSubmit={handleSubmit}
    />
  )
}
