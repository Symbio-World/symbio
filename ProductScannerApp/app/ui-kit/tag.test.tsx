import 'react-native'
import React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import { TouchableOpacity } from 'react-native'
import { Tag } from './tag'

describe('Tag', () => {
  it('renders correctly', () => {
    const { toJSON } = render(<Tag title='Test' />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('can be selected', () => {
    const { toJSON } = render(<Tag title='Test' selected />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('calls onPress', () => {
    const testOnPress = jest.fn()
    const { getByType } = render(<Tag title='Test' onPress={testOnPress} />)
    fireEvent.press(getByType(TouchableOpacity))
    expect(testOnPress).toHaveBeenCalled
  })
})
