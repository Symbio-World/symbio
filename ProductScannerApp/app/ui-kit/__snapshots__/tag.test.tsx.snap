// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag can be selected 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#48bb78",
      "borderRadius": 9999,
      "margin": 12,
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 20,
        },
        Object {
          "color": "#fff",
        },
        Object {
          "fontWeight": "bold",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;

exports[`Tag renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderRadius": 9999,
      "margin": 12,
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 12,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 20,
        },
        Object {
          "color": "#48bb78",
        },
        Object {
          "fontWeight": "bold",
        },
      ]
    }
  >
    Test
  </Text>
</View>
`;
